poppy/zernike.py

```diff
@@ -702,17 +702,42 @@ def arbitrary_basis(aperture, nterms=15, rho=None, theta=None):
     # code submitted by Arthur Vigan - see https://github.com/mperrin/poppy/issues/166
 
     shape = aperture.shape
-    npix  = shape[0]
-
-    A = aperture.sum()
+    assert len(shape) == 2 and shape[0] == shape[1], \
+        "only square aperture arrays are supported"
+
+    # To avoid clipping the aperture, we precompute the zernike modes
+    # on an array oversized s.t. the zernike disk circumscribes the 
+    # entire aperture. We then slice the zernike array down to the 
+    # requested array size and cut the aperture out of it.
+
+    # get max extent of aperture from array center
+    yind, xind = np.where(aperture > 0)
+    distance = np.sqrt( (yind-shape[0]/2.)**2 + (xind-shape[1]/2.)**2 )
```

Comment:
This is somewhat subtle but you actually want `yind-(shape[0]-1)/2` here, and likewise for x. Consider::

``` python
In [6]: shape=(3,3)

In [7]: yind,xind=np.indices(shape)

In [8]: yind-shape[0]/2.
Out[8]:
array([[-1.5, -1.5, -1.5],
       [-0.5, -0.5, -0.5],
       [ 0.5,  0.5,  0.5]])

In [9]: yind-(shape[0]-1)/2.
Out[9]:
array([[-1., -1., -1.],
       [ 0.,  0.,  0.],
       [ 1.,  1.,  1.]])

```